edge-modules/MetricsCollector/MetricsScrapeAndUpload.cs

```diff
@@ -38,18 +41,44 @@ public void Dispose()
 
         async Task ScrapeAndUploadMetricsAsync(CancellationToken cancellationToken)
         {
+            IEnumerable<Metric> scrapedMetrics = Enumerable.Empty<Metric>();
             try
             {
-                IEnumerable<Metric> metrics = await this.scraper.ScrapeEndpointsAsync(cancellationToken);
-                await this.publisher.PublishAsync(this.GetGuidTaggedMetrics(metrics), cancellationToken);
+                await this.PublishPreviouslyFailedMetricsAsync(cancellationToken);
+
+                scrapedMetrics = await this.scraper.ScrapeEndpointsAsync(cancellationToken);
+                scrapedMetrics = this.GetGuidTaggedMetrics(scrapedMetrics);
+                await this.publisher.PublishAsync(scrapedMetrics, cancellationToken);
                 Logger.LogInformation("Successfully scraped and uploaded metrics");
             }
             catch (Exception e)
             {
                 Logger.LogError(e, "Error scraping and uploading metrics");
+                this.failedUploadMetrics.AddRange(scrapedMetrics);
```

Comment:
publish logic of publisher is done by batch, right? For example. when exception happens at batch x; then batch 1 to (x-1) is sent and should not resend again.  Metrics in batch x or later should be added to failed metrics and will resend later on.  If my understanding is correct, you should not add all scraped metrics to failed upload metrics.